File: python/leetcode_bak/427_Construct_Quad_Tree.py
# /usr/bin/python3.6
# -*- coding:utf-8 -*-


# Definition for a QuadTree node.
class Node(object):
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight


class Solution(object):
    def judge(self, grid, start, end):
        ret = (True, grid[start][start])
        for i in range(start, end+1):
            for j in range(start, end+1):
                if grid[i][j] != ret[-1]:
                    return False, 0
        return ret

    def buffer(self,root, grid, x1, y1, x2, y2):
        mid_from_top_to_down = int((y1+y2)/2)
        mid_from_left_to_right = int((x1+x2)/2)
        # 区域划分
        # TopLeft(start, mid)

        # TopRight(
        # BottomLeft
        # BottomRight



    def construct(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: Node
        """


def main():
    s = Solution()


if __name__ == "__main__":
    main()

File: C++/LeetCode/LeetCode/427_Construct_Quad_Tree.cpp
#include<vector>
using namespace std;
class Node {
public:
	bool val;
	bool isLeaf;
	Node* topLeft;
	Node* topRight;
	Node* bottomLeft;
	Node* bottomRight;

	Node() {}

	Node(bool _val, bool _isLeaf, Node* _topLeft, Node* _topRight, Node* _bottomLeft, Node* _bottomRight) {
		val = _val;
		isLeaf = _isLeaf;
		topLeft = _topLeft;
		topRight = _topRight;
		bottomLeft = _bottomLeft;
		bottomRight = _bottomRight;
	}
};
class Solution {
public:
	bool check(vector<vector<int>>& grid, int left, int right, int top, int down) {
		bool flag = grid[top][left] == 1;
		for (int i = top; i <= down; i++) {
			for (int j = left; j <= right; j++) {
				if (flag != (grid[i][j] == 1)) {
					return false;
				}
			}
		}
		return true;
	}
	Node* buff(vector<vector<int>>& grid, int left, int right, int top, int down) {
		if (check(grid, left, right, top, down)) {
			return new Node(grid[top][left] == 1, true, NULL, NULL, NULL, NULL);
		}
		else {
			Node* cur = new Node(false, false, NULL, NULL, NULL, NULL);
			int middle_of_left_right = (left + right) / 2;
			int middle_of_top_down = (top + down) / 2;
			cur->topLeft = buff(grid, left, middle_of_left_right, top, middle_of_top_down);
			cur->topRight = buff(grid, middle_of_left_right + 1, right, top, middle_of_top_down);

			cur->bottomLeft = buff(grid, left, middle_of_left_right, middle_of_top_down + 1, down);
			cur->bottomRight = buff(grid, middle_of_left_right + 1, right, middle_of_top_down + 1, down);
			return cur;
		}
	}
	Node* construct(vector<vector<int>>& grid) {
		return buff(grid, 0, grid.size() - 1, 0, grid[0].size() - 1);
	}
};



